File: nodejs/bulkoperations/src/bulk-operations.js
const {CacheManager,CacheItem,JsonDataType} = require('ncache-client');
let cache = null;
const config = require('./app.config.json');
async function Run() {
    // Initialize cache 
    await InitializeCache();
    // Create a simple product object
    const products = CreateNewProducts();
    const keys = GetKeys(products);

    // Generate hashmap of the keys and products
    let productItems = GetCacheItemMap(keys, products);
    // Adding item
    await AddMultipleObjectsToCache(productItems);

    // Modify the object and update in cache
    await UpdateMultipleObjectsInCache(keys, productItems);

    // Get the objects from cache
    await GetMultipleObjectsFromCache(keys);

    // Delete the existing object
    await RemoveMultipleObjectsFromCache(keys);

    //close cache
    await cache.close();
}

async function InitializeCache() {
    // Get Cache Name
    const cacheName = config.cacheProperties.cacheName
    if (!cacheName) {
        console.log("The CacheID cannot be null or empty.");
        return;
    }
    // Initialize an instance of the cache to begin performing operations:
    cache = await CacheManager.getCache(cacheName);

    // Print output on console
    console.log(`Cache ${cacheName} is initialized.`);
            
}

async function AddMultipleObjectsToCache(items) {
 
    const result = await cache.addBulk(items);
    if (result.size === 0) {
        console.log("All items are successfully added to cache.");
    } else {
        console.log("One or more items could not be added to the cache. Iterate map for details.");
        for (let value of result.values()) {
            console.log(value)
          }
    }
}

async function GetMultipleObjectsFromCache(keys) {
    const items = await cache.getBulk(keys,JsonDataType.Object);
    if (items.size > 0)
    {
         // Print output on console
         console.log("Following items are fetched from cache.");
         for(let item of items.values()) {
            PrintProductDetails(item);
         }
    }
}
//change
async function UpdateMultipleObjectsInCache(keys, items) {
     // Update item
     const oldItemValue = items.get(keys[0]).getValue(JsonDataType.Object);
     oldItemValue.ClassName = "ClassC";
     items.get(keys[0]).setValue(oldItemValue,"product");

     const results = await cache.insertBulk(items);
     if (results.size === 0) {
        console.log("nAll items are successfully updated in cache.")
     } else {
        console.log("One or more items could not be added to the cache. Iterate hashmap for details.")
        for(let result of results.values()) {
            console.log(result);
         }
     }
}

async function RemoveMultipleObjectsFromCache(key) {
    // Remove the existing customer
    const items = await cache.removeBulk(key,JsonDataType.Object);
    if (items.size === 0) {
        console.log("nNo items removed from the cache against the provided keys.")
    } else {
        console.log("Following items have been removed from the cache:")
        for(let item of items.values()) {
            PrintProductDetails(item);
         }
    }
}

function GetKeys(products) {
    return products.map(product =>`Product:${product.Name}`);
}
function GetCacheItemMap(keys,products) {
    const items = new Map();
    for (i = 0 ; i < keys.length ; i++) {
        const cacheItem = new CacheItem(products[i],"product")
        items.set(keys[i],cacheItem);
    }
    return items;
}
function CreateNewProducts() {
    return([
        {Id : 1, Name : "Dairy Milk Cheese", ClassName : "ClassA", Category : "Edibles"},
        {Id : 2, Name : "American Butter", ClassName : "ClassA", Category : "Edibles"},
        {Id : 3, Name : "Walmart Delicious Cream", ClassName : "ClassA", Category : "Edibles"},
        {Id : 4, Name : "Nestle Yogurt", ClassName : "ClassA", Category : "Edibles"},
    ]);
}
function PrintProductDetails(product) {
    if (product == null) return;
    console.log("Id:       " + product.Id);
    console.log("Name:     " + product.Name);
    console.log("Class:    " + product.ClassName);
    console.log("Category: " + product.Category);
    console.log();

}
module.exports = {
    Run
}

File: python/bulk_operations/bulk_operations.py
from ncache.client.CacheManager import CacheManager
from ncache.client.CacheItem import CacheItem
from sample_data.product import Product


def initialize_cache():
    cache_name = "demoCache"
    cache = CacheManager.get_cache(cache_name)
    print("\nCache initialized successfully.")
    return cache


def create_products():
    product_1 = Product(id="1001", name="Tea", unit_price=50, units_available=20)
    product_2 = Product(id="1002", name="Milk", unit_price=30, units_available=30)
    product_3 = Product(id="1003", name="Butter", unit_price=30, units_available=20)
    product_4 = Product(id="1004", name="Cheese", unit_price=40, units_available=10)
    products = [product_1, product_2, product_3, product_4]
    return products


def get_keys(products):
    keys = []
    for product in products:
        keys.append("Product:" + product.id)
    return keys


def create_dictionary(keys, products):
    items = {}
    for i in range(len(products)):
        item = CacheItem(products[i])
        items[keys[i]] = item
    return items


def add_multiple_objects_in_cache(cache, items):
    keys_failed = cache.add_bulk(items)
    if len(keys_failed) == 0:
        print("\nAll items are added in cache.")
    else:
        print("\nOne or more items could not be added in cache. The items failed are:")
        for key in keys_failed.keys():
            print("\n", key)


def update_multiple_objects_in_cache(cache, keys, items):
    for i in range(len(items)):
        item = items[keys[i]]
        product = item.get_value(Product)
        product.unit_price = 100
        items[keys[i]] = CacheItem(product)

    keys_failed = cache.insert_bulk(items)
    if len(keys_failed) == 0:
        print("\nAll items are updated in cache.")
    else:
        print("\nOne or more items could not be updated in cache. The items failed are:")
        for key in keys_failed.keys():
            print("\n", key)


def get_multiple_objects_from_cache(cache, keys):
    fetched_products = cache.get_bulk(keys, Product)
    if len(fetched_products) != 0:
        for key in fetched_products:
            product = fetched_products[key]
            print(product)


def remove_multiple_objects_from_cache(cache, keys):
    removed_items = cache.remove_bulk(keys, Product)
    if len(removed_items) == len(keys):
        print("\nAll items are removed from cache.")
    else:
        print("\nOne or more keys failed to remove.")


def dispose_cache(cache):
    cache.close()
    print("\nCache closed.")


# Initialize instance of the cache to perform operations
cache = initialize_cache()

# Create multiple Product objects
products = create_products()
keys = get_keys(products)

# Create dictionary of keys and products
items = create_dictionary(keys, products)

# Adding multiple items in cache
add_multiple_objects_in_cache(cache, items)

# Update multiple items in cache
update_multiple_objects_in_cache(cache, keys, items)

# Get multiple items from cache
get_multiple_objects_from_cache(cache, keys)

# Remove the existing objects
remove_multiple_objects_from_cache(cache, keys)

# Dispose the cache once done
dispose_cache(cache)
